### PR TITLE
Add signer-keys service

### PR DESCRIPTION
**Pull request by snazy**
Adds a service to provide secret keys that can be used to sign and verify signatures. Keys are rotated automatically and shared across all Nessie instances.

This code is going to be used to sign and verify the Nessie S3 sign endpoint URIs.

### CODE REVIEW

File: catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/SignerSignature.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import jakarta.validation.constraints.NotNull;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import org.projectnessie.catalog.service.objtypes.SignerKey;
+import org.projectnessie.nessie.immutables.NessieImmutable;
+
+@NessieImmutable
+public abstract class SignerSignature {
```
Comment (snazy) [inline catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/SignerSignature.java:32]: This object is going to encapsulate all parameters for S3 signing, respecting writeable + read-only locations, object-storage-friendly write location and the signature.

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKeysObj.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    return OBJ_ID;
+  }
+
+  @Override
+  @Value.Default
+  default ObjType type() {
+    return OBJ_TYPE;
+  }
+
+  List<SignerKey> signerKeys();
```
Comment (adutra) [inline catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKeysObj.java:58]: Suggestion: turn this into `Map<String, SignerKey>` for faster key access in the method below.

File: catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+/**
+ * Signer keys service implementation based on {@link Persist} using a {@link SignerKeysObj} per
+ * Nessie repository. Relies on Nessie's distributed cache invalidation, when used in a horizontally
+ * scaled setup (multiple Nessie processes serving the same repository).
+ */
+@RequestScoped
+public class SignerKeysServiceImpl implements SignerKeysService {
+  public static final long SPIN_LOOP_MIN_SLEEP_MILLIS = 20;
+  public static final long SPIN_LOOP_MAX_SLEEP_MILLIS = 200;
+  public static final Duration NEW_KEY_ROTATE_AFTER = Duration.of(3, DAYS);
```
Comment (adutra) [inline catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java:50]: Should this be configurable?

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import org.immutables.value.Value;
+import org.projectnessie.nessie.immutables.NessieImmutable;
+
+@NessieImmutable
+@JsonSerialize(as = ImmutableSignerKey.class)
+@JsonDeserialize(as = ImmutableSignerKey.class)
+public interface SignerKey {
+  String name();
+
+  String secretKey();
```
Comment (adutra) [inline catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java:35]:
```suggestion
  /** Base64-encoded secret key. */
  String secretKey();
```

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  Instant creationTime();
+
+  Instant rotationTime();
+
+  Instant expirationTime();
+
+  @JsonIgnore
+  @Value.Lazy
+  default SecretKeySpec secretKeySpec() {
+    byte[] secretKeyBytes = ("NessieSign:" + secretKey()).getBytes(UTF_8);
```
Comment (adutra) [inline catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java:46]: This looks suspicious, shouldn't we decode `secretKey()` (it's base64-encoded) first? And why prepend `"NessieSign:"`?

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    return new SecretKeySpec(secretKeyBytes, "hmacSha256");
```
Comment (adutra) [inline catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java:47]: Technically, this key is not a "hmacSha256" key. It's more a "custom" algorithm.

File: catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/SignerSignature.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+  public abstract long expirationTimestamp();
+
+  @SuppressWarnings("UnstableApiUsage")
+  public String sign(SignerKey signerKey) {
+    Hasher hasher = hmacSha256(signerKey.secretKeySpec()).newHasher();
+    hasher.putLong(expirationTimestamp());
+    hasher.putString("prefix=" + prefix(), UTF_8);
+    hasher.putString("identifier=" + identifier(), UTF_8);
+    hasher.putString("b=" + warehouseLocation(), UTF_8);
```
Comment (adutra) [inline catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/SignerSignature.java:51]: Why `b`?

File: catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            persist.fetchTypedObj(
+                SignerKeysObj.OBJ_ID, SignerKeysObj.OBJ_TYPE, SignerKeysObj.class);
+        break;
+      } catch (ObjNotFoundException notFound) {
+        signerKeys = addNewKey(null);
+        try {
+          if (storeInitial(signerKeys)) {
+            break;
+          }
+          persistSpinLoop();
```
[the remark this answers is not in the document]
Comment (snazy) [reply catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java:76]: Race w/ another Nessie instance storing the obj

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKeysObj.java (hunk 2 above)
Comment (snazy) [reply catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKeysObj.java:58]: The access would be faster - deserialization, which happens every time the object's needed, slower.
The implementation also relies on the order of the elements (oldest first, newest last).

File: catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java (hunk 3 above)
Comment (snazy) [reply catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java:50]: Yea - want to do that as a follow-up after #9170

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java (hunk 4 above)
Comment (snazy) [reply catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java:35]: Actually, this should be a byte[]

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java (hunk 6 above)
Comment (snazy) [reply catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java:47]: damn, right

File: catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/SignerSignature.java (hunk 7 above)
Comment (snazy) [reply catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/SignerSignature.java:51]: for "bucket" - "w" is already used for _writeable locations_

File: catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (adutra) [reply catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java:75]: Yes, but now that this instance lost the insertion battle, why wait to fetch the object that the other instance inserted?

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKeysObj.java (hunk 2 above)
Comment (adutra) [reply catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKeysObj.java:58]: Good point, maybe make this a derived/lazy attribute?

File: catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java (hunk 6 above)
Comment (adutra) [reply catalog/service/common/src/main/java/org/projectnessie/catalog/service/objtypes/SignerKey.java:51]: The key is going to serve as the base key for a hmacSha256 encryption later on, but the key itself was generated by a custom algorithm. But it doesn't matter, it seems the value passed here is not relevant and is never read.

File: catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (snazy) [reply catalog/service/impl/src/main/java/org/projectnessie/catalog/service/impl/SignerKeysServiceImpl.java:75]: True, removed

**Review by adutra — APPROVED** [review]

**Pull request merged**